File: web/src/api/dispatch.ts
import { http } from "@/utils/request";

export interface DispatchOrder {
  id: number;
  project_id: number | null;
  source_type: string;
  source_id: number | null;
  title: string;
  root_cause_hint: string | null;
  level: string | null;
  status: string;
  assignee_id: number | null;
  assignee_name: string | null;
  deadline: string | null;
  description: string | null;
  last_action_note: string | null;
  closed_at: string | null;
  created_by: number | null;
  created_at: string | null;
  updated_at: string | null;
}

export interface DispatchPage {
  total: number;
  items: DispatchOrder[];
  page: number;
  size: number;
}

export interface DispatchStats {
  total: number;
  by_status: Record<string, number>;
  by_level: Record<string, number>;
}

export interface DispatchOptions {
  statuses: string[];
  sources: string[];
  levels: string[];
}

export interface DispatchPreset {
  source_type: string;
  source_id?: number | null;
  project_id?: number | null;
  title?: string;
  root_cause_hint?: string | null;
  level?: string | null;
}

export function getDispatchOptions(): Promise<DispatchOptions> {
  return http<DispatchOptions>({ url: "/v1/dispatch/options", method: "GET" });
}

export function getDispatchStats(): Promise<DispatchStats> {
  return http<DispatchStats>({ url: "/v1/dispatch/stats", method: "GET" });
}

export function listDispatches(params: {
  status?: string;
  source_type?: string;
  project_id?: number;
  page?: number;
  size?: number;
}): Promise<DispatchPage> {
  return http<DispatchPage>({ url: "/v1/dispatch/", method: "GET", params });
}

export function getDispatch(id: number): Promise<DispatchOrder> {
  return http<DispatchOrder>({ url: `/v1/dispatch/${id}`, method: "GET" });
}

export function createDispatch(data: {
  title: string;
  source_type?: string;
  source_id?: number | null;
  project_id?: number | null;
  level?: string | null;
  root_cause_hint?: string | null;
  assignee_id?: number | null;
  deadline?: string | null;
  description?: string | null;
}): Promise<DispatchOrder> {
  return http<DispatchOrder>({ url: "/v1/dispatch/", method: "POST", data });
}

export function dispatchAction(id: number, action: string, note?: string): Promise<DispatchOrder> {
  return http<DispatchOrder>({ url: `/v1/dispatch/${id}`, method: "PATCH", data: { action, note } });
}

export function reassignDispatch(id: number, assignee_id: number, note?: string): Promise<DispatchOrder> {
  return http<DispatchOrder>({ url: `/v1/dispatch/${id}/reassign`, method: "POST", data: { assignee_id, note } });
}

File: app/schema/dispatch.py
"""根因派单闭环 Schema（#80）：创建/动作流转/改派/详情/分页/统计。

时间字段统一以「北京时间墙钟」字符串对外（YYYY-MM-DDTHH:mm:ss），与前端
el-date-picker 的 value-format 对齐；入库存 timestamptz（约定按 Asia/Shanghai 解释）。
"""

from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field, field_serializer

from app.core.clock import LOCAL_TZ
from app.core.constants import (
    DISPATCH_LEVELS,
    DISPATCH_SOURCES,
    DISPATCH_STATUSES,
)


class DispatchCreate(BaseModel):
    title: str = Field(..., description="派单标题")
    source_type: str = Field("manual", description="来源(correlation/alarm/manual)")
    source_id: int | None = Field(None, description="来源记录ID(事件组/告警)")
    project_id: int | None = Field(None, description="归属项目(manual 来源必填，其余由来源解析)")
    level: str | None = Field(None, description="级别(严重/警告/提示)")
    root_cause_hint: str | None = Field(None, description="根因提示")
    assignee_id: int | None = Field(None, description="处理人ID")
    deadline: datetime | None = Field(None, description="处理时限")
    description: str | None = Field(None, description="处置说明/要求")


class DispatchAction(BaseModel):
    """状态机动作：start(待派→处理中) / close(处理中→已闭环) / reopen(已闭环→处理中)。"""

    action: str = Field(..., description="动作(start/close/reopen)")
    note: str | None = Field(None, description="处置备注")


class DispatchReassign(BaseModel):
    assignee_id: int = Field(..., description="新处理人ID")
    note: str | None = Field(None, description="改派备注")


class DispatchOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    project_id: int | None = None
    source_type: str
    source_id: int | None = None
    title: str
    root_cause_hint: str | None = None
    level: str | None = None
    status: str
    assignee_id: int | None = None
    assignee_name: str | None = None
    deadline: datetime | None = None
    description: str | None = None
    last_action_note: str | None = None
    closed_at: datetime | None = None
    created_by: int | None = None
    created_at: datetime | None = None
    updated_at: datetime | None = None

    @field_serializer("deadline", "closed_at", "created_at", "updated_at")
    def _serialize_dt(self, v: datetime | None) -> str | None:
        if v is None:
            return None
        return v.astimezone(LOCAL_TZ).replace(tzinfo=None).isoformat()


class DispatchPage(BaseModel):
    total: int
    items: list[DispatchOut]
    page: int = 1
    size: int = 20


class DispatchStats(BaseModel):
    total: int
    by_status: dict[str, int]
    by_level: dict[str, int]


# 供前端下拉复用的枚举
DISPATCH_STATUS_OPTIONS = list(DISPATCH_STATUSES)
DISPATCH_SOURCE_OPTIONS = list(DISPATCH_SOURCES)
DISPATCH_LEVEL_OPTIONS = list(DISPATCH_LEVELS)
